sklearn/ensemble/iforest.py

```diff
@@ -0,0 +1,239 @@
+            # ensemble sorts the indices.
+            X.sort_indices()
+
+        rnd = check_random_state(self.random_state)
+        y = rnd.uniform(size=X.shape[0])
+        self.max_samples = min(int(X.shape[0]), self.max_samples)
+        super(IsolationForest, self).fit(X, y, sample_weight=sample_weight)
+        return self
+
+    def _cost(self, n):
+        """ The average path length in a n samples iTree, which is equal to
+        the average path length of an unsuccessful BST search since the
+        latter has the same structure as an isolation tree.
+        """
+        if n <= 1:
+            return 1.
+        else:
+            harmonic_number = np.log(n) + 0.5772156649
+            return 2. * harmonic_number - 2. * (n - 1.) / n
+
+    def predict(self, X):
+        # code structure from ForestClassifier/predict_proba
+        """Predict anomaly score of X with the IsolationForest algorithm.
+
+        The anomaly score of an input sample is computed as
+        the mean anomaly scores of the trees in the forest.
+
+        The measure of normality of an observation given a tree is the depth
+        of the leaf containing this observation, which is equivalent to
+        the number of splitting required to isolate this point. In case of
+        several observations n_left in the leaf, the average lenght path of
+        a n_left samples isolation tree is added.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix of shape = [n_samples, n_features]
+            The input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
+            to a sparse ``csr_matrix``.
+
+        Returns
+        -------
+        scores : array of shape = [n_samples]
+            The anomaly score of the input samples.
+            The higher, the more abnormal.
+        """
+        # Check data
+        X = check_array(X, dtype=DTYPE, accept_sparse="csr")
+        n_samples = X.shape[0]
+
+        # Assign chunk of trees to jobs
+        n_jobs, n_trees, starts = _partition_estimators(self.n_estimators,
+                                                        self.n_jobs)
+
+        # Parallel loop
+        results = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,
+                           backend="threading")(
+            delayed(_parallel_helper)(tree.tree_, 'apply_depth', X)
+            for tree in self.estimators_)
+
+        # Reduce
+        results = np.array(results)
+        depth = np.zeros(n_samples)
+        scores = np.zeros(n_samples)
+
+        depth = np.mean(results, axis=0)
+
+        for k in range(n_samples):
+            scores[k] = np.power(2, - depth[k] / self._cost(n_samples))
+
+        return scores
+
+    def decision_function(self, X):
+        """Average of the decision functions of the base classifiers.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape = [n_samples, n_features]
```

Comment:
{array-like, sparse matrix} of shape = [n_samples, n_features]
->
{array-like, sparse matrix}, shape (n_samples, n_features)